File: fjlc/preprocessing/preprocessors/TweetNGramsPMI.java
package com.freva.masteroppgave.preprocessing.preprocessors;

import com.freva.masteroppgave.preprocessing.filters.Filters;
import com.freva.masteroppgave.preprocessing.filters.RegexFilters;
import com.freva.masteroppgave.classifier.ClassifierOptions;
import com.freva.masteroppgave.utils.progressbar.Progressable;
import com.freva.masteroppgave.utils.reader.LineReader;
import com.freva.masteroppgave.utils.tools.Parallel;

import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.concurrent.atomic.AtomicInteger;


public class TweetNGramsPMI implements Progressable {
    private LineReader tweetReader;
    private NGramTree nGramTree;

    /**
     * Finds all frequent (and meaningful) n-grams in a file, treating each new line as a new document.
     *
     * @param input        LineReader initialized on file with documents to generate n-grams for
     * @param n            Maximum n-gram length
     * @param minFrequency Smallest required frequency to include n-gram
     * @param minPMI       Minimum PMI value for n-gram to be included
     * @param filters      List of filters to apply to document before generating n-grams
     * @return Map of n-grams as key and number of occurrences as value
     */
    public final List<String> getFrequentNGrams(LineReader input, int n, double minFrequency, double minPMI, Filters filters) {
        final AtomicInteger lineCounter = new AtomicInteger(0);
        tweetReader = input;
        nGramTree = new NGramTree();

        Parallel.For(tweetReader, tweet -> {
            synchronized (lineCounter) {
                if (lineCounter.incrementAndGet() % 200000 == 0) {
                    nGramTree.pruneInfrequent((int) Math.ceil(minFrequency * lineCounter.intValue() / 2));
                }
            }

            tweet = filters.apply(tweet);
            for (String sentence : RegexFilters.SENTENCE_END_PUNCTUATION.split(tweet)) {
                String[] tokens = RegexFilters.WHITESPACE.split(sentence.trim());
                if (tokens.length == 1) continue;

                for (int i = 0; i < tokens.length; i++) {
                    nGramTree.incrementNGram(Arrays.copyOfRange(tokens, i, Math.min(i + n, tokens.length)));
                }
            }
        });

        return nGramTree.getNGrams((int) (minFrequency * lineCounter.intValue()), minPMI);
    }


    @Override
    public double getProgress() {
        return tweetReader != null ? tweetReader.getProgress() : 0;
    }


    private class NGramTree {
        private Node root = new Node("");

        private synchronized void incrementNGram(String[] nGram) {
            Node current = root;
            current.numOccurrences++;

            for (String word : nGram) {
                if (!current.hasChild(word)) {
                    current.addChild(word);
                }

                current = current.getChild(word);
                current.numOccurrences++;
            }
        }

        private Node getNode(String phrase) {
            Node current = root;
            for (String word : RegexFilters.WHITESPACE.split(phrase)) {
                if (!current.hasChild(word)) {
                    return null;
                }

                current = current.getChild(word);
            }
            return current;
        }

        private void pruneInfrequent(int limit) {
            root.pruneInfrequent(limit);
        }

        private List<String> getNGrams(int limit, double inclusionThreshold) {
            Map<String, Double> allNGrams = new HashMap<>();

            for (Node child : root.children.values()) {
                child.addFrequentPhrases(allNGrams, limit, child.phrase);
            }

            List<String> filteredNGrams = new ArrayList<>();
            for (Map.Entry<String, Double> next : allNGrams.entrySet()) {
                String[] nGramTokens = RegexFilters.WHITESPACE.split(next.getKey());

                if (next.getValue() >= inclusionThreshold &&
                        !ClassifierOptions.containsIntensifier(nGramTokens) &&
                        !ClassifierOptions.isStopWord(nGramTokens[nGramTokens.length - 1])) {
                    filteredNGrams.add(next.getKey());
                }
            }

            return filteredNGrams;
        }
    }


    private class Node {
        private Map<String, Node> children = new HashMap<>();
        private String phrase;
        private int numOccurrences;
        private double logScore = Double.NaN;

        public Node(String phrase) {
            this.phrase = phrase;
        }

        public boolean hasChild(String value) {
            return children.containsKey(value);
        }

        public void addChild(String value) {
            children.put(value, new Node(value));
        }

        public Node getChild(String value) {
            return children.get(value);
        }

        public double getLogScore() {
            if (Double.isNaN(logScore)) logScore = Math.log(numOccurrences);
            return logScore;
        }

        private void pruneInfrequent(int limit) {
            Iterator<Map.Entry<String, Node>> iterator = children.entrySet().iterator();
            while (iterator.hasNext()) {
                Map.Entry<String, Node> child = iterator.next();

                if (child.getValue().numOccurrences < limit) {
                    iterator.remove();
                } else {
                    child.getValue().pruneInfrequent(limit);
                }
            }
        }

        private void addFrequentPhrases(Map<String, Double> map, int limit, String prefix) {
            children.values().stream()
                    .filter(child -> child.numOccurrences >= limit)
                    .forEach(child -> {
                        Node lastWord = nGramTree.getNode(child.phrase);

                        if (lastWord != null && lastWord.numOccurrences >= limit) {
                            double temp = nGramTree.root.getLogScore() + child.getLogScore() - getLogScore() - lastWord.getLogScore();

                            String candidate = prefix + " " + child.phrase;
                            map.put(candidate, temp);
                            child.addFrequentPhrases(map, limit, candidate);
                        }
                    });
        }
    }
}

File: fjlc/preprocessing/preprocessors/tweet_n_grams_pmi.py
import math

from fjlc.classifier import classifier_options
from fjlc.preprocessing.filters.regex_filters import RegexFilters


class TweetNGramsPMI:

    tweet_reader = None
    n_gram_tree = None

    def get_frequent_n_grams(self, input_reader, n, min_frequency, min_pmi, filters):
        """
        Finds all frequent (and meaningful) n-grams in a file, treating each new line as a new document.
        
        :param input_reader:    LineReader initialized on file with documents to generate n-grams for
        :param n:               Maximum n-gram length
        :param min_frequency:   Smallest required frequency to include n-gram
        :param min_pmi:         Minimum PMI value for n-gram to be included
        :param filters:         List of filters to apply to document before generating n-grams
        :return:                Map of n-grams as key and number of occurrences as value
        """
        line_counter = 0
        TweetNGramsPMI.tweet_reader = input_reader
        TweetNGramsPMI.n_gram_tree = self.NGramTree()

        # Todo: Parallelize
        for tweet in self.tweet_reader:
            line_counter += 1
            if line_counter % 200000 == 0:
                TweetNGramsPMI.n_gram_tree.prune_infrequent(math.ceil(min_frequency * line_counter / 2.))

            tweet = filters.apply(tweet)
            for sentence in RegexFilters.SENTENCE_END_PUNCTUATION.split(tweet):
                tokens = RegexFilters.WHITESPACE.split(sentence.strip())
                if len(tokens) == 1:
                    continue

                for i in range(len(tokens)):
                    self.n_gram_tree.increment_n_gram(tokens[i:min(i + n, len(tokens))])

        return self.n_gram_tree.get_n_grams(int(min_frequency * line_counter), min_pmi)

    @staticmethod
    def get_progress():
        return 0 if TweetNGramsPMI.tweet_reader is None else TweetNGramsPMI.tweet_reader.get_progress()

    class NGramTree:

        def __init__(self):
            self.root = TweetNGramsPMI.Node("")

        def increment_n_gram(self, n_gram):
            current = self.root
            current.num_occurrences += 1

            for word in n_gram:
                if not current.has_child(word):
                    current.add_child(word)

                current = current.get_child(word)
                current.num_occurrences += 1

        def get_node(self, phrase):
            current = self.root
            for word in RegexFilters.WHITESPACE.split(phrase):
                if not current.has_child(word):
                    return None

                current = current.get_child(word)
            return current

        def prune_infrequent(self, limit):
            self.root.prune_infrequent(limit)

        def get_n_grams(self, limit, inclusion_threshold):
            all_n_grams = {}

            for child in self.root.children.values():
                child.add_frequent_phrases(all_n_grams, limit, child.phrase)

            filtered_n_grams = []
            for next_key, next_value in all_n_grams.items():
                n_gram_tokens = RegexFilters.WHITESPACE.split(next_key)

                if next_value >= inclusion_threshold and not classifier_options.contains_intensifier(n_gram_tokens) and not classifier_options.is_stop_word(n_gram_tokens[-1]):
                    filtered_n_grams.append(next_key)

            return filtered_n_grams

    class Node:
        def __init__(self, phrase):
            self.children = {}
            self.phrase = phrase
            self.num_occurrences = 0
            self.log_score = 0.0

        def has_child(self, value):
            return value in self.children

        def add_child(self, value):
            self.children[value] = TweetNGramsPMI.Node(value)

        def get_child(self, value):
            return self.children[value]

        def get_log_score(self):
            if self.log_score == 0.0:
                self.log_score = math.log(self.num_occurrences)
            return self.log_score

        def prune_infrequent(self, limit):
            self.children = [child for child in self.children if child.get_value().num_occurrences >= limit]
            for child in self.children:
                child.get_value().prune_infrequent(limit)

        def add_frequent_phrases(self, dictionary, limit, prefix):
            for child in self.children.values():
                if child.num_occurrences < limit:
                    continue
                last_word = TweetNGramsPMI.n_gram_tree.get_node(child.phrase)

                if last_word is not None and last_word.num_occurrences >= limit:
                    temp = TweetNGramsPMI.n_gram_tree.root.get_log_score() + child.get_log_score() - self.get_log_score() - last_word.get_log_score()

                    candidate = prefix + " " + child.phrase
                    dictionary[candidate] = temp
                    child.add_frequent_phrases(dictionary, limit, candidate)
